jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Session.java

```diff
@@ -675,26 +674,54 @@ public boolean close(int error, String reason, Callback callback)
         while (true)
         {
             CloseState current = closed.get();
-            switch (current)
+            if (current == CloseState.NOT_CLOSED)
             {
-                case NOT_CLOSED:
+                if (closed.compareAndSet(current, CloseState.LOCALLY_CLOSED))
                 {
-                    if (closed.compareAndSet(current, CloseState.LOCALLY_CLOSED))
-                    {
-                        closeFrame = newGoAwayFrame(CloseState.LOCALLY_CLOSED, error, reason);
-                        control(null, callback, closeFrame);
-                        return true;
-                    }
-                    break;
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("Closing {}/{}", error, reason);
+                    closeFrame = newGoAwayFrame(CloseState.LOCALLY_CLOSED, error, reason);
+                    control(null, callback, closeFrame);
+                    return true;
                 }
-                default:
+            }
+            else
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("Ignoring close {}/{}, already closed", error, reason);
+                callback.succeeded();
+                return false;
+            }
+        }
+    }
+
+    @Override
+    public CompletableFuture<Void> shutdown()
```

Comment:
I would clearly document that this `shutdown` method has to be idempotent as `HTTP2SessionContainer` might call it twice, and explain the circumstances under which it can be called twice.